File: app/core/history.py
import copy
from typing import List, Callable
import numpy as np
from app.utils.logger import logger

class Command:
    """Encapsulates an undoable operation."""
    def execute(self):
        pass

    def undo(self):
        pass

class MaskEditCommand(Command):
    """Command for mask modifications (Brush, Eraser, Auto Remove, Magic Wand, Filters)."""
    def __init__(self, document, old_mask: np.ndarray, new_mask: np.ndarray, description: str = "Edit Mask"):
        self.document = document
        self.old_mask = old_mask.copy() if old_mask is not None else None
        self.new_mask = new_mask.copy() if new_mask is not None else None
        self.description = description

    def execute(self):
        if self.new_mask is not None:
            self.document.mask = self.new_mask.copy()
            self.document.notify_changed()

    def undo(self):
        if self.old_mask is not None:
            self.document.mask = self.old_mask.copy()
            self.document.notify_changed()

class UndoStack:
    """Manages command history stack for Undo / Redo functionality."""

    def __init__(self, max_depth: int = 30):
        self.max_depth = max_depth
        self._undo_stack: List[Command] = []
        self._redo_stack: List[Command] = []
        self._on_change_callbacks: List[Callable] = []

    def push(self, command: Command):
        command.execute()
        self._undo_stack.append(command)
        if len(self._undo_stack) > self.max_depth:
            self._undo_stack.pop(0)
        self._redo_stack.clear()
        self._notify()

    def undo(self):
        if not self.can_undo():
            return
        cmd = self._undo_stack.pop()
        cmd.undo()
        self._redo_stack.append(cmd)
        self._notify()

    def redo(self):
        if not self.can_redo():
            return
        cmd = self._redo_stack.pop()
        cmd.execute()
        self._undo_stack.append(cmd)
        self._notify()

    def can_undo(self) -> bool:
        return len(self._undo_stack) > 0

    def can_redo(self) -> bool:
        return len(self._redo_stack) > 0

    def clear(self):
        self._undo_stack.clear()
        self._redo_stack.clear()
        self._notify()

    def add_change_listener(self, callback: Callable):
        self._on_change_callbacks.append(callback)

    def _notify(self):
        for cb in self._on_change_callbacks:
            try:
                cb()
            except Exception as e:
                logger.error(f"Error in UndoStack change listener: {e}")

File: Cpp/src/core/History.cpp
#include "core/History.h"
#include "core/ImageDocument.h"
#include "utils/Logger.h"

namespace ImageCut {
namespace Core {

// MaskEditCommand
MaskEditCommand::MaskEditCommand(ImageDocument* document,
                                 const cv::Mat& oldMask,
                                 const cv::Mat& newMask,
                                 const QString& layerId,
                                 const QString& description)
    : m_document(document), m_layerId(layerId), m_description(description)
{
    if (!oldMask.empty()) m_oldMask = oldMask.clone();
    if (!newMask.empty()) m_newMask = newMask.clone();
}

void MaskEditCommand::execute() {
    if (!m_document) return;
    auto target = m_layerId.isEmpty() ? m_document->getActiveLayer() : m_document->getLayerById(m_layerId);
    if (target) {
        target->mask = m_newMask.empty() ? cv::Mat() : m_newMask.clone();
        m_document->notifyChanged();
    }
}

void MaskEditCommand::undo() {
    if (!m_document) return;
    auto target = m_layerId.isEmpty() ? m_document->getActiveLayer() : m_document->getLayerById(m_layerId);
    if (target) {
        target->mask = m_oldMask.empty() ? cv::Mat() : m_oldMask.clone();
        m_document->notifyChanged();
    }
}

// DocumentActionCommand
DocumentActionCommand::DocumentActionCommand(ImageDocument* document,
                                             std::function<void()> undoFn,
                                             std::function<void()> redoFn,
                                             const QString& description)
    : m_document(document), m_undoFn(undoFn), m_redoFn(redoFn), m_description(description)
{}

void DocumentActionCommand::execute() {
    if (m_redoFn) m_redoFn();
    if (m_document) m_document->notifyChanged();
}

void DocumentActionCommand::undo() {
    if (m_undoFn) m_undoFn();
    if (m_document) m_document->notifyChanged();
}

// UndoStack
UndoStack::UndoStack(size_t maxDepth) : m_maxDepth(maxDepth) {}

void UndoStack::push(std::unique_ptr<Command> command) {
    if (!command) return;
    command->execute();
    m_undoStack.push_back(std::move(command));

    if (m_undoStack.size() > m_maxDepth) {
        m_undoStack.erase(m_undoStack.begin());
    }
    m_redoStack.clear();
    notify();
}

void UndoStack::undo() {
    if (!canUndo()) return;
    auto cmd = std::move(m_undoStack.back());
    m_undoStack.pop_back();

    cmd->undo();
    m_redoStack.push_back(std::move(cmd));
    notify();
}

void UndoStack::redo() {
    if (!canRedo()) return;
    auto cmd = std::move(m_redoStack.back());
    m_redoStack.pop_back();

    cmd->execute();
    m_undoStack.push_back(std::move(cmd));
    notify();
}

bool UndoStack::canUndo() const {
    return !m_undoStack.empty();
}

bool UndoStack::canRedo() const {
    return !m_redoStack.empty();
}

void UndoStack::clear() {
    m_undoStack.clear();
    m_redoStack.clear();
    notify();
}

void UndoStack::addChangeListener(std::function<void()> callback) {
    m_onChangeCallbacks.push_back(callback);
}

void UndoStack::notify() {
    for (auto& cb : m_onChangeCallbacks) {
        if (cb) {
            try { cb(); } catch (...) {}
        }
    }
}

} // namespace Core
} // namespace ImageCut
